feat: add support for nested transaction rollbacks via savepoints in sql

LucianBuzzo opened the pull request:
~~This adds a high level test that should pass if/when https://github.com/prisma/prisma-engines/pull/4375 is merged.~~
Transactions are now handled inside this repository, so the update to prisma-engines is no longer required.

This change adds support for handling rollbacks in nested transactions
in SQL databases. Specifically, the inner transaction should be rolled
back if the outer transaction fails.

To do this we keep track of the transaction ID and transaction depth so we can
re-use an existing open transaction in the underlying engine. This change also
allows the use of the `$transaction` method on an interactive transaction client.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Nested interactive transactions for SQL databases (Postgres, MySQL, SQLite, SQL Server) via savepoints; MongoDB remains without nested-transaction support.

* **Tests**
  * Added extensive nested-transaction tests covering multi-level nesting, savepoint behavior, rollback/commit propagation, concurrency and provider-specific syntax.

* **Documentation**
  * Updated runtime documentation to describe nested transaction behavior and provider-specific savepoint semantics.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LucianBuzzo committed:
feat: Add support for nested transaction rollbacks in SQL databases

This change adds support for handling rollbacks in nested transactions
in SQL databases. Specifically, the inner transaction should be rolled
back if the outer transaction fails.

To do this we keep track of the transaction ID and transaction depth so we can
re-use an existing open transaction in the underlying engine. This change also
allows the use of the `$transaction` method on an interactive transaction client.

depends-on: https://github.com/prisma/prisma-engines/pull/4375

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=21678) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=21678) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=21678) it.</sub>

Jolg42 commented on packages/client/tests/functional/interactive-transactions/tests.ts:
```diff
@@ -166,6 +166,46 @@ testMatrix.setupTestSuite(({ provider, providerFlavor }, _suiteMeta, clientMeta)
+   * If a parent transaction is rolled back, the child transaction should also rollback
+   * - This is only supported in SQL derived servers
+   */
+  testIf(provider.toLowerCase().includes('sql'))('sql: nested rollback', async () => {
```
```suggestion
  testIf(provider !== Providers.MONGODB)('sql: nested rollback', async () => {
```

Jolg42 replied:
Please add this import along this change
```
import { Providers } from '../_utils/providers'
```

LucianBuzzo commented:
@Jolg42 Is it possible to verify that https://github.com/prisma/prisma-engines/pull/4375 fixes the issue and gets the test to pass?

janpio commented:
`5.6.0-12.integration-sql-nested-transactions-a8af640da343c10e0f7421264c2838b302d3ce6d` via https://github.com/prisma/prisma/pull/21735/files should work for that.

janpio commented:
Ummm, seems it is not working @LucianBuzzo. Or I messed up royally. Asking some colleagues for advice.

LucianBuzzo commented:
@janpio I'm replicating the issue on my end - I think I might need to update the client code to maintain the same transaction ID in nested transactions 🤔 . I'm going to dig through and see what I can come up wiht.

Jolg42 commented on docker/docker-compose.yml:
```diff
@@ -60,17 +60,17 @@ services:
       timeout: 2s
       retries: 20
 
-  cockroachdb:
```
You'll need to undo the changes done here below to make tests pass in GitHub Actions.

janpio commented:
Replaced by https://github.com/prisma/prisma/pull/22161

janpio commented on packages/engines/package.json:
```diff
@@ -22,7 +22,7 @@
   "dependencies": {
     "@prisma/debug": "workspace:*",
-    "@prisma/engines-version": "5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a",
+    "@prisma/engines-version": "5.7.0-22.integration-sql-nested-transactions2-f486ca20b160ea1fd92f8f351f6d05eff8b478b3",
```
```suggestion
    "@prisma/engines-version": "5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8",
```

janpio commented on packages/fetch-engine/package.json:
```diff
@@ -43,7 +43,7 @@
   "dependencies": {
     "@prisma/debug": "workspace:*",
-    "@prisma/engines-version": "5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a",
+    "@prisma/engines-version": "5.7.0-22.integration-sql-nested-transactions2-f486ca20b160ea1fd92f8f351f6d05eff8b478b3",
```
```suggestion
    "@prisma/engines-version": "5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8",
```

janpio commented on packages/internals/package.json:
```diff
@@ -81,7 +81,7 @@
     "@prisma/generator-helper": "workspace:*",
     "@prisma/get-platform": "workspace:*",
-    "@prisma/prisma-schema-wasm": "5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a",
+    "@prisma/prisma-schema-wasm": "5.7.0-22.integration-sql-nested-transactions2-f486ca20b160ea1fd92f8f351f6d05eff8b478b3",
```
```suggestion
    "@prisma/prisma-schema-wasm": "5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8",
```

janpio commented on packages/migrate/package.json:
```diff
@@ -56,7 +56,7 @@
   "dependencies": {
     "@prisma/debug": "workspace:*",
-    "@prisma/engines-version": "5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a",
+    "@prisma/engines-version": "5.7.0-22.integration-sql-nested-transactions2-f486ca20b160ea1fd92f8f351f6d05eff8b478b3",
```
```suggestion
    "@prisma/engines-version": "5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8",
```

janpio commented on pnpm-lock.yaml:
```diff
@@ -467,8 +467,8 @@ importers:
       '@prisma/engines-version':
-        specifier: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
-        version: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
+        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
+        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```
```suggestion
        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```

janpio commented on pnpm-lock.yaml:
```diff
@@ -763,8 +763,8 @@ importers:
       '@prisma/engines-version':
-        specifier: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
-        version: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
+        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
+        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```
```suggestion
        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```

janpio commented on pnpm-lock.yaml:
```diff
@@ -1088,8 +1088,8 @@ importers:
       '@prisma/prisma-schema-wasm':
-        specifier: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
-        version: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
+        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
+        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```
```suggestion
        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```

janpio commented on pnpm-lock.yaml:
```diff
@@ -1245,8 +1245,8 @@ importers:
       '@prisma/engines-version':
-        specifier: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
-        version: 5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a
+        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
+        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```
```suggestion
        specifier: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
        version: 5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8
```

janpio commented on pnpm-lock.yaml:
```diff
@@ -3166,17 +3166,17 @@ packages:
 
-  /@prisma/prisma-schema-wasm@5.7.0-21.f2e5886144b9dcd9774a6799317e94d76904001a:
-    resolution: {integrity: sha512-cWUGImR3ztSJpb2z+Vka/IS6rFEAvQ6AHlcH92btu+iAjcPpc46YlI23iujvfBpZySJALJmxdji+CNVG2ybWMg==}
+  /@prisma/prisma-schema-wasm@5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8:
+    resolution: {integrity: sha512-/bUANcYz7Zfkks62e/qWZscn369bqfPWmx86GkQ8AJ7HuJmds8lkOOOQ8vr7zPGvbHfn00SrVXtabLmQdfk0fQ==}
```
```suggestion
  /@prisma/prisma-schema-wasm@5.7.0-29.integration-sql-nested-transactions2-e943433a36b16524e1d88f0f0b93c3d8d4ddd4c8:
    resolution: {integrity: sha512-/bUANcYz7Zfkks62e/qWZscn369bqfPWmx86GkQ8AJ7HuJmds8lkOOOQ8vr7zPGvbHfn00SrVXtabLmQdfk0fQ==}

```

LucianBuzzo commented:
@janpio The tests are passing locally. I'm now working on an edge case where you use a batch transaction inside a client extensions 😬

Jolg42 commented:
@LucianBuzzo to get the CI running, you would need to solve the git conflict. GitHub won't run any Action if there is a conflict with the base branch unfortunately 😢

LucianBuzzo commented:
@Jolg42 Cool, I'll get that sorted 👍

LucianBuzzo commented:
@Jolg42 I'm getting a lot of tests failing with an error about `relationJoins`, e.g.
```
issues.16535-select-enum (provider=mysql) › allows to select enum field

    GetDmmfError: Prisma schema validation - (get-dmmf wasm)
    Error code: P1012
    error: The preview feature "relationJoins" is not known. Expected one of: deno, driverAdapters, fullTextIndex, fullTextSearch, metrics, multiSchema, postgresqlExtensions, tracing, views
      -->  schema.prisma:6
       | 
     5 | engineType = "library"
     6 | previewFeatures = ["relationJoins"]
       | 
       
       ```
       
       Any idea why this is happening?

Jolg42 commented:
@LucianBuzzo 
>EDIT: I'm going to guess that the engine version I'm using doesn't have this feature - I'll work on getting the engine PR updated

Exactly, the `main` evolves rapidly, so getting your fork in sync would work I think.

LucianBuzzo commented:
The new interactive rollback tests are now passing and behaving as expected, however there are now some failing tests elsewhere that I'll need to address.

LucianBuzzo commented:
From what I can tell the transaction depth isn't getting decremented when using the `js_pg` adapter.
I've added a test to sanity check this:
```
 pnpm run test:functional --adapter js_pg --shard 6/6 --client-runtime wasm --preview-features driverAdapters --runInBand -t 'multiple interactive transactions'
 ```

LucianBuzzo commented:
@SevInf It looks like everything is now passing, except d1 and mini-proxy, which looks to be unreleated.

SevInf commented:
`mini-proxy` might actually be related: did we change binary-engine public interface in the engine PR?
Both `mini-proxy` and real Accelerate use it underneath, so we might take a closer look to ensure they won't break once it released. Of course, both of those projects are in the private repos, so there is nothing really actionable for you right now. I'll try to carve out some time to take a look myself, but realistically, that probably happen after 5.12.0 release next week.

LucianBuzzo commented:
@SevInf I've added a `begin()` method in the engine that is used to open the next transaction level on a transaction that has already been started. That might be the culprit. If there's nothing I can do to help, I'll sit tight until after the v5.12.0 release 👍

LucianBuzzo commented:
@SevInf Now that v5.12.0 is released, are you able to assist in the mini-proxy and d1 test failures? TIA 🙏

LucianBuzzo commented:
@SevInf @janpio Any updates on how to move this PR forwards?

janpio commented:
@LucianBuzzo I contacted you via LinkedIn (only method I could find) to talk about how getting a better way for us to talk to each other (Either invite you to our Slack, or via Discord?)

I merged, which of course lead to a few more tests failing - [`417db07` (#21678)](https://github.com/prisma/prisma/pull/21678/commits/417db0742e0f5b3fa0fa55dff3cdb3344adcd365) is the last commit before I did that to get to the test failures that probably exist and are real, and not just caused by new tests that the engine version here does not support yet.

LucianBuzzo commented:
@janpio Thanks, I'll check out the test failures - I've responded to you message on LinkedIn, slack is probably the best bet 👍

LucianBuzzo commented:
@SevInf Looks like the mini-proxy tests were failing because I was accidentally overwriting the transaction payload when making a request to the engine. This is fixed and everything appears to be working correctly. 
I think the `Client func&legacy-notypes` might be flakey - can you rerun the failed test?

aqrln commented on packages/adapter-neon/src/neon.ts:
```diff
@@ -154,6 +154,13 @@ class NeonTransaction extends NeonWsQueryable<neon.PoolClient> implements Transa
     super(client)
   }
 
+  async begin(): Promise<Result<void>> {
+    debug(`[js::begin]`)
+
+    this.client.release()
+    return Promise.resolve(ok(undefined))
+  }
```
I'm not sure I understand this. Why does `begin` close the transaction?

LucianBuzzo replied:
This is a mistake, I think maybe a copypasta from the `commit` implementation 🤔 
I'll remove the release line.

aqrln commented on packages/client/src/runtime/core/types/exported/itxClientDenyList.ts:
```diff
@@ -1,4 +1,4 @@
-const denylist = ['$connect', '$disconnect', '$on', '$transaction', '$use', '$extends'] as const
+const denylist = ['$connect', '$disconnect', '$on', '$use', '$extends'] as const
```
Shouldn't this depend on the provider now?

LucianBuzzo replied:
Yes correct, this is really only available for SQL servers. Do you know if there is a utility for determining if the provider is an SQL server?

LucianBuzzo committed:
fix: disallow concurrent nested transactions

For now, concurrent nested transactions are not allowed. With the
current architecture, concurrent nested transactions would disrupt the
depth/savepoint tracking in the engine, resulting in unexpected
behaviour. As a future feature, a queuing system or more complex
savepoint tracking could be introduced to support the concurrency.

aqrln commented on packages/client/tests/functional/extensions/itx.ts:
```diff
@@ -315,7 +315,7 @@ testMatrix.setupTestSuite(
               expect(ctx.$connect).toBeUndefined()
               expect(ctx.$disconnect).toBeUndefined()
-              expect(ctx.$transaction).toBeUndefined()
+              expect(ctx.$transaction).toBeDefined()
```
This should be undefined on MongoDB

LucianBuzzo replied:
@aqrln My preference would be to leave it defined so there is a consistent interface, and throw a readable error, like what is done for the planetscale adapter.

LucianBuzzo committed:
feat: add test to check that nested transactions are unavailable on MongoDB

LucianBuzzo committed:
fix: add begin() to benchmark mock transaction

LucianBuzzo committed:
fix(transaction): enforce nesting order for itx

LucianBuzzo committed:
fix(transaction): block closing itx while nested scope active

LucianBuzzo committed:
test(itx): cover nested commit + no partial writes

LucianBuzzo committed:
test(itx): add sequential/deep nesting + catch inner rollback

LucianBuzzo committed:
fix(itx): scope tracking per transaction + force rollback on order violations

LucianBuzzo committed:
feat(client-engine-runtime): add savepoints for nested transactions

LucianBuzzo committed:
test(itx): add nested transaction edge cases + mongodb type/runtime coverage

LucianBuzzo committed:
Fix savepoint rollback syntax; drop unused Transaction.begin

LucianBuzzo committed:
Fix itx payload merge; keep savepoint state consistent

LucianBuzzo committed:
fix(lint): ignore .generated and remove unnecessary async in tests

LucianBuzzo committed:
chore: remove whitespace-only diffs

LucianBuzzo committed:
Merge branch 'prisma:main' into lucianbuzzo/nested-rollbacks

LucianBuzzo committed:
fix linter issues

jacek-prisma commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -672,46 +678,110 @@ Or read our docs at https://www.prisma.io/docs/concepts/components/prisma-client
         isolationLevel: options?.isolationLevel ?? this._engineConfig.transactionOptions.isolationLevel,
+        newTxId: options.newTxId,
+      }
+      let info: any
```
Why not use the actual type? `InteractiveTransactionInfo`

LucianBuzzo committed:
fix: serialize nested transaction state transitions

Summary:

- Added per-transaction operation serialization in TransactionManager via operationQueue and lock helpers, so nested start/commit/rollback and timeout/cancel paths cannot interleave state mutations.

- Routed nested start (newTxId) through the transaction lock and revalidated state under lock before mutating depth/savepoint stacks.

- Updated nested rollback to keep state consistent even when rollback-to-savepoint or release-savepoint queries fail.

- Added a regression test for concurrent nested starts where the first savepoint creation fails, verifying the second nested start remains consistent and rollback targets the correct savepoint.

- Removed leftover RequestHandler interactiveTransaction payload mutation that injected new_tx_id.

- Updated AGENTS.md touched line to use official Wasm capitalization.

Validation:

- pnpm --filter @prisma/client-engine-runtime test transaction-manager.test.ts

- pnpm --filter @prisma/client-engine-runtime build

LucianBuzzo committed:
refactor: tighten nested itx scope typing in getPrismaClient

Changes:

- Added getItxScopeContext() helper to read TX_ID/TX_SCOPE_ID/TX_SCOPE_STATE with a typed context object, removing repeated local any-casts.

- Updated _transactionWithCallback() to use typed itx context values for nested transaction scope handling.

- Pulled scopeStack.push(scopeId) out of the nested/non-nested if-else without changing behavior.

- Replaced transaction start info type from any to Transaction.InteractiveTransactionInfo<unknown>.

- Renamed rollbackAttempts to rollbackScopeCount for clearer intent in rollback loop.

- Replaced scopeStack.splice(0, scopeStack.length) with scopeStack.length = 0 for direct stack reset.

- Switched Mongo nested-transaction runtime guard to use typed itx context (txId presence) instead of direct symbol access.

jacek-prisma commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -216,6 +216,27 @@ type EventCallback<E extends ExtendedEventType> = [E] extends ['beforeExit']
+  txId: string | undefined
+  scopeId: string | undefined
+  scopeState: ItxScopeState | undefined
+}
```
I think these should always either all be undefined or all defined, the type should model these 2 states instead of handling each field separately

LucianBuzzo committed:
refactor: apply remaining nested itx cleanup feedback

Changes:

- Standardized provider checks in client-generator-ts to use context.isSqlProvider() for interactive transaction deny-list generation.

- Refined getPrismaClient transaction scope typing into coherent top-level vs nested states to avoid per-field optional handling.

- Added helpers for scope context validation and nested scope id generation to simplify _transactionWithCallback flow.

- Simplified nested-vs-top-level close order checks for readability while preserving existing runtime behavior and error messages.

- Added debug logging for suppressed rollback failures in multi-rollback cleanup loop.

- Updated Mongo nested transaction runtime check to use typed scope context kind.

LucianBuzzo committed:
docs: clarify per-transaction serialization guards

Added rationale comments above #withActiveTransactionLock and #runSerialized explaining that transaction operations are serialized per tx id to avoid savepoint/depth/status races across await boundaries.

LucianBuzzo committed:
refactor: move savepoint SQL ownership toward adapters

- Add optional Transaction.savepoint(action, name) API in @prisma/driver-adapter-utils and re-export SavepointAction.

- Update binder to forward wrapped savepoint methods for error-capturing adapters.

- Refactor TransactionManager nested start/commit/rollback to prefer adapter-provided savepoint SQL, with provider-based fallback retained behind a TODO.

- Implement savepoint methods for pg, neon, ppg, planetscale, mariadb, libsql, better-sqlite3, and mssql transaction implementations.

- Extend transaction-manager tests to verify adapter-provided savepoint SQL and optional release behavior.

- Validation: transaction-manager test suite passes; touched packages type-check with tsconfig.build.json using --emitDeclarationOnly false --noEmit.

jacek-prisma commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -216,6 +216,56 @@ type EventCallback<E extends ExtendedEventType> = [E] extends ['beforeExit']
+    return false
+  }
+
+  return Array.isArray((value as Record<string, unknown>).stack)
```
```suggestion
  return Array.isArray(value['stack'])
```

LucianBuzzo committed:
refactor: move nested savepoints to adapter async methods

- Replace Transaction savepoint(action, name) with async createSavepoint/rollbackToSavepoint/releaseSavepoint in driver-adapter-utils and binder.

- Refactor TransactionManager nested transaction flow to call adapter savepoint methods and remove provider fallback SQL synthesis.

- Migrate adapter savepoint implementations (pg, neon, ppg, planetscale, mariadb, libsql, better-sqlite3, mssql, d1-http, d1-worker) to execute async savepoint operations.

- Update transaction-manager tests for method-based savepoint behavior and missing-method error cases.

- Simplify nested ITX scope symbol storage in getPrismaClient to TX_SCOPE_CONTEXT.

- Document the new Transaction savepoint API shape in AGENTS.md.

jacek-prisma committed:
fix: add missing newTxId in schema

jacek-prisma committed:
chore: remove cast

jacek-prisma committed:
fix: update QPE handling

jacek-prisma committed:
Merge branch 'main' into lucianbuzzo/nested-rollbacks

jacek-prisma committed:
Merge branch 'main' into lucianbuzzo/nested-rollbacks
